code/hud/hudtargetbox.cpp

```diff
@@ -1566,43 +1566,65 @@ void get_turret_subsys_name(ship_weapon *swp, char *outstr)
 
 	//WMC - find the first weapon, if there is one
 	if (swp->num_primary_banks || swp->num_secondary_banks) {
-		auto flags = turret_weapon_aggregate_flags(swp);
-
-		// check if beam or flak using weapon flags
-		if (flags[Weapon::Info_Flags::Beam]) {
-			sprintf(outstr, "%s", XSTR("Beam turret", 1567));
-		} else if (flags[Weapon::Info_Flags::Flak]) {
-			sprintf(outstr, "%s", XSTR("Flak turret", 1566));
-		} else {
-			if (turret_weapon_has_subtype(swp, WP_MISSILE)) {
-				sprintf(outstr, "%s", XSTR("Missile lnchr", 1569));
-			} else if (turret_weapon_has_subtype(swp, WP_LASER)) {
-				// ballistic too! - Goober5000
-				if (flags[Weapon::Info_Flags::Ballistic])
-				{
-					sprintf(outstr, "%s", XSTR("Turret", 1487));
-				}
-				// the TVWP has some primaries flagged as bombs
-				else if (flags[Weapon::Info_Flags::Bomb])
-				{
-					sprintf(outstr, "%s", XSTR("Missile lnchr", 1569));
+		bool done = false;
+		if (swp->num_primary_banks) {
+			for (auto weaponid : swp->primary_bank_weapons) {
+				if (weaponid >= 0) {
```

Comment:
Why not loop from 0 to `swp->num_primary_banks`?  All weaponids after that point are going to be -1 anyway.